### ISSUE TITLE
WhiteRabbit availability on Maven?

### ISSUE THREAD

**Issue by passionfruit18** [OHDSI/WhiteRabbit#156]
Hi,

We are trying to use WhiteRabbit as part of a Grails project and were wondering if it could be made available on a Maven repository for ease of use as a dependency.

Thanks
James

**Comment by clairblacketer** [comment]
Hi @passionfruit18 I believe with the [most recent release](https://github.com/OHDSI/WhiteRabbit/releases/tag/v0.8.0) the build was switched from eclipse to maven

**Comment by passionfruit18** [comment]
Hi @clairblacketer , if I understand correctly, in the most recent release, White Rabbit is using Maven to handle its own dependencies. But this does not necessarily mean White Rabbit itself is available on a Maven repository to be used as a dependency by other Maven projects?

**Comment by clairblacketer** [comment]
@passionfruit18 I am admittedly out of my depth at this point. Can @schuemie or possibly @blootsvoets help answer the question?

**Comment by blootsvoets** [comment]
It would make sense to publish version 0.8.0 to a maven repo. Indeed, we could configure Maven `pom.xml` to upload do that. However, I don't have credentials to do that personally. Example code from the circe repo:
```xml
<project ...>
        <repositories>
            <repository>
              <id>ohdsi</id>
              <name>repo.ohdsi.org</name>
              <url>http://repo.ohdsi.org:8085/nexus/content/repositories/releases</url>
            </repository>
            <repository>
              <id>ohdsi.thirdparty</id>
              <name>repo.ohdsi.org</name>
              <url>http://repo.ohdsi.org:8085/nexus/content/repositories/thirdparty</url>
            </repository>
            <repository>
              <id>ohdsi.snapshots</id>
              <name>repo.ohdsi.org-snapshots</name>
              <url>http://repo.ohdsi.org:8085/nexus/content/repositories/snapshots</url>
              <releases>
                <enabled>false</enabled>
              </releases>
              <snapshots>
                <enabled>true</enabled>
              </snapshots>
            </repository>
        </repositories>
        ...
</project>
```
It would make sense to upload the dependencies in the `lib/` folder to the `ohdsi.thirdparty` repository, to remove all local dependencies. Perhaps a developer from circe or webapi can help out here?

**Comment by schuemie** [comment]
@leeevans, @chrisknoll : I think you've set this up for Circe? What would it take to get WhiteRabbit on Maven?

@passionfruit18 : I'm also trying to understand the use case here. WhiteRabbit and RiaH are both stand-alone Java applications. In what scenario would they serve as dependencies for other applications?

**Comment by passionfruit18** [comment]
Hi @schuemie , we have a workflow where we take the profile of some data and then use that profile in turn to construct a "Data Model". We would like to automate this process so that instead of manually using the White Rabbit application and producing an excel profile and then uploading the excel profile to our application, which constructs a Data Model, we simply upload data to our application, and it uses White Rabbit behind the scenes.

**Comment by schuemie** [comment]
Thanks @passionfruit18 for that clarification!

**Comment by chrisknoll** [comment]
Hi @schuemie  and @passionfruit18,

I think @leeevans  can set up an automated job to publish to maven repository.  It would be similar to CIRCE in that the job will build and push to nexus based on the version in pom.xml.  But I'll let @leeevans  speak to that.

As far as which nexus repo to put it into: I would put snapshot builds under `snapshots` and `releases` under releases.  I wouldn't put anything under `thirdparty`: that is used to host external libraries that OHDSI code might be depend on, but isn't code managed by OHDSI itself.  @schuemie , correct me if I'm wrong, but White Rabbit should be considered an OHDSI tool (since it is in the OHDSI organization) and so publishing to snapshot/releases I think is appropriate here.

@schuemie I was also confused by publishing an entire application to nexus (I think of nexus as a repository of shared libraries, not a place to publish applications) but there are some classes exposed by the White Rabbit application that could be referenced by other apps, so I can see the usecase here.  Ideally, I think you'd want to architect WR as an application that is assembled from different components (each with their own maven artifact) and then White Rabbit references those components, but then other applications can reference the same parts to handle their own use-cases. this would save the bandwidth of downloading all the WR application/UI code that they may not be using for their `behind the scenes` work.  But, theoretically, WR is just a JAR with classes in it, so ti import it as a maven dependency should allow access to those classes.

**Comment by schuemie** [comment]
Thanks @chrisknoll ! Yes, I would also consider WR an OHDSI tool, so `snapshot` and `releases` sound correct (but @leeevans  gets the final word).

I agree a reorganization of WR would be ideal, but it is not a very high priority for me. If someone else wants to pick that up they are more than welcome to it.

**Comment by blootsvoets** [comment]
Specifically the files in `https://github.com/OHDSI/WhiteRabbit/tree/master/lib/com/teradata` (not under our control) would be nice if we can submit them to `thirdparty`. Indeed the main code can go to `releases` and `snapshots`.

While building the maven files, I already split out a `rabbit-core` module, which contains code referenced by both WhiteRabbit and Rabbit In A Hat. Ideally `rabbit-core` would be a dependency that is also usable in external code. @passionfruit18 can you verify what code of [`rabbit-core`](https://github.com/OHDSI/WhiteRabbit/tree/master/rabbit-core/src/main/java/org/ohdsi) and [White  Rabbit](https://github.com/OHDSI/WhiteRabbit/tree/master/whiterabbit/src/main/java/org/ohdsi/whiteRabbit) you are using? That would help identifying libraries useful for external developers versus UI application code.

**Comment by leeevans** [comment]
@schuemie @chrisknoll @blootsvoets Yes, if you can setup the WR pom.xml with the references to the OHDSI nexus repos in a similar manner to the CIRCE pom.xml, I can help with creating a Jenkins job that automates publishing the artifacts to the OHDSI nexus snapshot and released repos.

**Comment by passionfruit18** [comment]
@blootsvoets Actually it is a colleague who is mainly responsible for using it, I have linked him to this discussion so he should provide the details soon.

**Comment by johncross** [comment]
Hi @blootsvoets , I'm using org.ohdsi.whiteRabbit.DbSettings and org.ohdsi.whiteRabbit.scan.SourceDataScan.process() which are both in WhiteRabbit - I'm not using any of rabbit-core directly at the moment. This is still very early stages though, still testing!

**Comment by chrisknoll** [comment]
> Specifically the files in https://github.com/OHDSI/WhiteRabbit/tree/master/lib/com/teradata (not under our control) would be nice if we can submit them to thirdparty.

Right,t hat makes sense, those were the 'local dependencies' that eclipse probably used to load local JARs, but in maven world, they should be hosted somewhere in a nexus repo.  We don't have to drop them into `thirdparty` if they already have a published location (like maven central). If there's any libraries that aren't discoverable (ie, not published) then that's where we'd put them in thirdparty (subject to third party terms of licensing of course).
